xgcm/grid.py

```diff
@@ -2247,7 +2251,67 @@ def cumsum(self, da, axis, **kwargs):
 
         >>> grid.max(da, ["X", "Y"], fill_value={"X": 0, "Y": 100})
         """
-        return self._1d_grid_ufunc_dispatch("cumsum", da, axis, **kwargs)
+
+        boundary_width = {}
+
+        if isinstance(axis, str):
+            axis = [axis]
+        to = self._as_axis_kwarg_mapping(to)
+
+        axes = [self.axes[ax_name] for ax_name in axis]
+        for ax in axes:
+            pos, dim = ax._get_position_name(da)
+
+            # first use xarray's cumsum method
+            # TODO if we knew all the dims we could do this in one go instead of looping
+            da_cum = da.cumsum(dim=dim)
+
+            ax_to = to[ax.name]
+            if ax_to is None:
+                ax_to = ax._default_shifts[pos]
+
+            # now pad / trim the data as necessary
+            # here we enumerate all the valid possible shifts
+            if (pos == "center" and ax_to == "right") or (
+                pos == "left" and ax_to == "center"
+            ):
+                # do nothing, this is the default for how cumsum works
+                data = da_cum
+                boundary_width[ax.name] = (0, 0)
+            elif (pos == "center" and ax_to == "left") or (
+                pos == "right" and ax_to == "center"
+            ):
+                data = da_cum.isel(**{dim: slice(0, -1)})
+                boundary_width[ax.name] = (1, 0)
+            elif (pos == "center" and ax_to == "inner") or (
+                pos == "outer" and ax_to == "center"
+            ):
+                data = da_cum.isel(**{dim: slice(0, -1)})
+                boundary_width[ax.name] = (0, 0)
+            elif (pos == "center" and ax_to == "outer") or (
+                pos == "inner" and ax_to == "center"
+            ):
+                data = da_cum
+                boundary_width[ax.name] = (1, 0)
+            else:
+                raise ValueError(
+                    f"From `{pos}` to `{ax_to}` is not a valid position "
+                    f"shift for cumsum operation along axis {ax}."
+                )
+
+        data = pad(
+            data=data,
+            grid=self,
+            boundary_width=boundary_width,
+            boundary=boundary,
+            fill_value=fill_value,
+        )
+
+        # TODO need to rename dims to their new positions here for this to work
+
+        return _reattach_coords(
+            [data], grid=self, boundary_width=boundary_width, keep_coords=True
+        )
```

Comment:
Does this return a list of data_arrays by any chance? 
I am seeing many errors of [this type ](https://github.com/xgcm/xgcm/actions/runs/3485376148/jobs/5830857086#step:6:62292) in the tests.